keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,435 @@
+Because currently (as of k8s-1.14) the scale down is only limited by [Stabilization Window][].
+In order to repeat the default behavior we set `constraints.scaleDown.delaySeconds` to 5min
+(the default value for Stabilization window), and let it rule the number of pods.
+
+[Stabilization Window]: #stabilization-window
+
+#### The Motivation To “Pick The Largest Constraint” Concept
+
+Take a look at the example:
+
+- `curReplicas = 10`
+- `calculatedReplicas = 20`
+
+The user specifies only one HPA parameter `constraints.scaleUp.pods = 5` and expects that number of replicas to be set to 15 during the next HPA controller loop.
+But the algorithm picks the largest change instead:
+
+    Constraint1 = 10 * 2 = 20 (as constraints.scaleUp.percent = 100 by default)
+    Constraint2 = 10 + 5 = 15 (as constraints.scaleUp.pods = 5, set by the user)
+    scaleUpLimit = max(20, 15) = 20
+    desiredReplicas = 20
+
+The user might expect that the autoscaler would use the smallest constraint (15), not the largest one (20). This behavior is not intuitive, but it does make sense if considered thoroughly.
+
+The main idea of the HPA is to autoscale because of a load increase to avoid request failures. It should work on both small clusters and large ones. For small clusters, the absolute number constraint works better (ScaleUpPods), for large clusters the percentage works better (ScaleUpPercentage).
+
+Example: If the current cluster size is `1` and calculated cluster size for this particular load is `20`, then we should reach it ASAP.
+
+For default values (ScaleUpPercent = 100, ScaleUpPods = 4) and “pick the largest constraint” concept, we’ll increase 1 -> 20 in 3 steps
+
+    1 -> 5 -> 10 -> 20
+
+The first step will use “scaleUp.pods” constraint; the next steps will use “scaleUp.percent” one.
+
+In case of more intuitive “pick the hardest limit” concept, we’ll increase the cluster in 6 steps:
+
+    1 -> 2 -> 4 -> 8 -> 12 -> 16 -> 20
+
+Given that each step takes [90 sec in worst case], we’ll respond to the load increase in `(6-3)*90 sec = ~ 5 min`.
+
+That’s too much, we should respond faster than that.
+
+[90 sec in worst case]: https://dzone.com/articles/kubernetes-autoscaling-101-cluster-autoscaler-hori-1
+
+The scale down constraints are a method to prevent too rapid loss of capacity. Hence, it makes sense to pick the maximum of two constraints.
+
+#### Stabilization Window
+
+Currently stabilization window ([PR][], [RFC][], [Algorithm Details][]) is used to gather “scale-down-recommendations” during some time (default is 5min),
+and a new number of replicas is set to the maximum of all recommendations.
+
+It may be defined by command line option `--horizontal-pod-autoscaler-downscale-stabilization-window`.
+
+The same idea is used in the `DelaySeconds` parameter above:
+
+- While scaling up, we should pick the safest (largest) "desiredReplicas" number during last `delaySeconds`.
```

Comment:
Picking the largest makes this redundant, since that's what the `ScaleDown.delaySeconds` does.  I would expect this to choose the **smallest** number during `delaySeconds`.  (I know that's different from the rate up/down semantics both of which choose the max.)

`ScaleDown.delaySeconds` has prior art in the commandline flag and people have asked for this configuration.  So we should implement it.

`ScaleUp.delaySeconds` use cases are more esoteric.  If you want to punt on the implementation for now, that's fine.  At least we have a place for it.

Reply:
yep, that was just a typo.
I also added new command line options. I believe, it is a good idea.
And also made the "stabilization-window" option obsolete.

Could you check plz? https://github.com/kubernetes/enhancements/pull/883/commits/ec9c48e2a701c88b1afd88a204bc70a17aa9eb1f